b2sdk/transfer/emerge/executor.py

```diff
@@ -0,0 +1,480 @@
+    ):
+        self.session = session
+        self.services = services
+        self.bucket_id = bucket_id
+        self.file_name = file_name
+        self.content_type = content_type
+        self.file_info = file_info
+        self.progress_listener = progress_listener
+
+    @abstractmethod
+    def execute_plan(self, emerge_plan):
+        pass
+
+
+class SmallFileEmergeExecution(BaseEmergeExecution):
+    def execute_plan(self, emerge_plan):
+        emerge_parts = list(emerge_plan.emerge_parts)
+        assert len(emerge_parts) == 1
+        emerge_part = emerge_parts[0]
+        execution_step_factory = SmallFileEmergeExecutionStepFactory(self, emerge_part)
+        execution_step = execution_step_factory.get_execution_step()
+        future = execution_step.execute()
+        return future.result()
+
+
+class LargeFileEmergeExecution(BaseEmergeExecution):
+    MAX_LARGE_FILE_SIZE = 10 * 1000 * 1000 * 1000 * 1000  # 10 TB
+
+    def __init__(
+        self,
+        session,
+        services,
+        bucket_id,
+        file_name,
+        content_type,
+        file_info,
+        progress_listener,
+        continue_large_file_id=None,
+        max_queue_size=None,
+    ):
+        super(LargeFileEmergeExecution, self).__init__(
+            session, services, bucket_id, file_name, content_type, file_info, progress_listener
+        )
+        self.continue_large_file_id = continue_large_file_id
+        self.max_queue_size = max_queue_size
+
+    def execute_plan(self, emerge_plan):
+        total_length = emerge_plan.get_total_length()
+
+        if total_length is not None and self.MAX_LARGE_FILE_SIZE < total_length:
+            raise MaxFileSizeExceeded(total_length, self.MAX_LARGE_FILE_SIZE)
+
+        plan_id = emerge_plan.get_plan_id()
+
+        file_info = dict(self.file_info) if self.file_info is not None else {}
+        if plan_id is not None:
+            file_info['plan_id'] = plan_id
+
+        self.progress_listener.set_total_bytes(total_length or 0)
+
+        emerge_parts_dict = None
+        if total_length is not None:
+            emerge_parts_dict = dict(emerge_plan.enumerate_emerge_parts())
+
+        unfinished_file, finished_parts = self._get_unfinished_file_and_parts(
+            self.bucket_id,
+            self.file_name,
+            file_info,
+            self.continue_large_file_id,
+            emerge_parts_dict=emerge_parts_dict
+        )
+
+        if unfinished_file is None:
+            if self.content_type is None:
+                content_type = self.DEFAULT_CONTENT_TYPE
+            else:
+                content_type = self.content_type
+            unfinished_file = self.services.large_file.start_large_file(
+                self.bucket_id, self.file_name, content_type, file_info
+            )
+        file_id = unfinished_file.file_id
+
+        semaphore = None
+        if self.max_queue_size is not None:
+            semaphore = threading.Semaphore(self.max_queue_size)
```

Comment:
maybe we could use a null semaphore to avoid `if`s down the line?

Reply:
you mean some dummy semaphore class thats do nothing...? here it is explicit that there is a possible state when we don't protect from number of submitted Futures - if there would be a "null semaphore" you would need to read a line with semaphore creation to know about that... I don't like it

Reply:
You need to read the semaphore constuction line anyway. Null semaphore would be equivalent to a semaphore with an infinite limit (or `maxint`). The rest of the code would be less jumpy though.

Reply:
I don't like the idea - semaphore has to protect counter with mutex - so this is some additional unnecessary synchronization logic just for some easiness in the code... - I know that then one may make a bug by forgetting about that `semaphore` can me `None` - but I think people should carefully write/change the code...